ghcide/src/Development/IDE/Core/FileStore.hs

```diff
@@ -290,3 +309,43 @@ setSomethingModified state = do
     -- Update database to remove any files that might have been renamed/deleted
     atomically $ writeTQueue (indexQueue $ hiedbWriter $ shakeExtras state) deleteMissingRealFiles
     void $ shakeRestart state []
+
+registerFileWatches :: [String] -> LSP.LspT Config IO Bool
+registerFileWatches globs = do
+      watchSupported <- isWatchSupported
+      if watchSupported
+      then do
+        let
+          regParams    = LSP.RegistrationParams (List [LSP.SomeRegistration registration])
+          -- The registration ID is arbitrary and is only used in case we want to deregister (which we won't).
+          -- We could also use something like a random UUID, as some other servers do, but this works for
+          -- our purposes.
+          registration = LSP.Registration "globalFileWatches"
+                                           LSP.SWorkspaceDidChangeWatchedFiles
+                                           regOptions
+          regOptions =
+            DidChangeWatchedFilesRegistrationOptions { _watchers = List watchers }
+          -- See Note [File existence cache and LSP file watchers] for why this exists, and the choice of watch kind
+          watchKind = WatchKind { _watchCreate = True, _watchChange = True, _watchDelete = True}
+          -- See Note [Which files should we watch?] for an explanation of why the pattern is the way that it is
+          -- The patterns will be something like "**/.hs", i.e. "any number of directory segments,
+          -- followed by a file with an extension 'hs'.
+          watcher glob = FileSystemWatcher { _globPattern = glob, _kind = Just watchKind }
+          -- We use multiple watchers instead of one using '{}' because lsp-test doesn't
+          -- support that: https://github.com/bubba/lsp-test/issues/77
+          watchers = [ watcher (Text.pack glob) | glob <- globs ]
+
+        void $ LSP.sendRequest LSP.SClientRegisterCapability regParams (const $ pure ()) -- TODO handle response
```

Comment:
Ideally we would block waiting for the response return and return a result depending on whether it was successful.

Reply:
This is an old TODO, it's not something added in this PR. I agree that we should wait for the response, but maybe not for this PR

Reply:
BTW I checked this and lsp-test doesn't seem to implement this part of the protocol, opened https://github.com/haskell/lsp/issues/333